File: code/functionapp/ServiceBus.cs
using Azure.Messaging.ServiceBus;
using common;
using LanguageExt;
using System;
using System.Collections.Generic;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Threading;
using System.Threading.Tasks;

namespace functionapp;

public static class ServiceBusModule
{
    public static async ValueTask QueueVirtualMachineCreation(ServiceBusClient client, VirtualMachineCreationQueueName queueName, Seq<(VirtualMachine VirtualMachine, DateTimeOffset EnqueueAt)> virtualMachines, CancellationToken cancellationToken)
    {
        var sender = client.CreateSender(queueName.Value);
        var messages = virtualMachines.Map(item => CreateMessage(item.VirtualMachine, item.EnqueueAt));
        var queue = new Queue<ServiceBusMessage>(messages);

        while (queue.Count > 0)
        {
            using var batch = await sender.CreateMessageBatchAsync(cancellationToken);

            // Remove messages from queue and add them to batch until the batch is full
            while (queue.Count > 0 && batch.TryAddMessage(queue.Peek()))
            {
                queue.Dequeue();
            }

            if (batch.Count > 0)
            {
                await sender.SendMessagesAsync(batch, cancellationToken);
            }
            else
            {
                throw new InvalidOperationException("Could not add message to batch. Payload might be too big.");
            }
        };
    }

    public static async ValueTask QueueOctaneBenchmark(ServiceBusClient client, OctaneBenchmarkQueueName queueName, VirtualMachine virtualMachine, CancellationToken cancellationToken)
    {
        var message = CreateMessage(virtualMachine);

        await SendMessage(client, queueName.Value, message, cancellationToken);
    }

    public static async ValueTask QueueVirtualMachineDeletion(ServiceBusClient client, VirtualMachineDeletionQueueName queueName, VirtualMachineName virtualMachineName, CancellationToken cancellationToken)
    {
        var message = CreateMessage(virtualMachineName);

        await SendMessage(client, queueName.Value, message, cancellationToken);
    }

    private static ServiceBusMessage CreateMessage(VirtualMachine virtualMachine)
    {
        var jsonObject = Serialize(virtualMachine);

        return CreateMessage(jsonObject);
    }

    private static ServiceBusMessage CreateMessage(VirtualMachine virtualMachine, DateTimeOffset enqueueAt)
    {
        var jsonObject = Serialize(virtualMachine);

        return CreateMessage(jsonObject, enqueueAt);
    }

    private static ServiceBusMessage CreateMessage(VirtualMachineName virtualMachineName)
    {
        var jsonObject = Serialize(virtualMachineName);

        return CreateMessage(jsonObject);
    }

    private static JsonObject Serialize(VirtualMachine virtualMachine)
    {
        return new JsonObject()
        {
            ["virtualMachineName"] = virtualMachine.Name.ToString(),
            ["virtualMachineSku"] = virtualMachine.Sku.ToString()
        };
    }

    private static JsonObject Serialize(VirtualMachineName virtualMachineName)
    {
        return new JsonObject()
        {
            ["virtualMachineName"] = virtualMachineName.ToString()
        };
    }

    private static ServiceBusMessage CreateMessage(JsonObject jsonObject)
    {
        var messageBytes = JsonSerializer.SerializeToUtf8Bytes(jsonObject);

        return new ServiceBusMessage(messageBytes);
    }

    private static ServiceBusMessage CreateMessage(JsonObject jsonObject, DateTimeOffset enqueueAt)
    {
        var messageBytes = JsonSerializer.SerializeToUtf8Bytes(jsonObject);

        return new ServiceBusMessage(messageBytes)
        {
            ScheduledEnqueueTime = enqueueAt
        };
    }

    private static async ValueTask SendMessage(ServiceBusClient client, string queueName, ServiceBusMessage message, CancellationToken cancellationToken)
    {
        var sender = client.CreateSender(queueName);

        await sender.SendMessageAsync(message, cancellationToken);
    }
}

File: code/functionapp/ServiceBus.fs
namespace functionapp

open Azure.Messaging.ServiceBus
open FSharp.Control
open Microsoft.Azure.WebJobs.ServiceBus
open System
open System.Text.Json.Nodes

open common

type ServiceBusQueueName =
    private
    | ServiceBusQueueName of NonEmptyString

    static member fromString value =
        if String.IsNullOrWhiteSpace(value) then
            invalidOp "Virtual machine SKU cannot be empty."
        else
            value
            |> NonEmptyString.fromString
            |> ServiceBusQueueName

    static member toString(ServiceBusQueueName nonEmptyString) = NonEmptyString.toString nonEmptyString

[<RequireQualifiedAccess>]
module ServiceBusMessage =
    let fromJsonObject (jsonObject: JsonObject) =
        jsonObject
        |> JsonObject.toBytes
        |> ServiceBusMessage

    let scheduleFromJsonObject jsonObject enqueueAt =
        let message = fromJsonObject jsonObject
        message.ScheduledEnqueueTime <- enqueueAt
        message

[<RequireQualifiedAccess>]
module ServiceBusClient =
    let sendMessage (client: ServiceBusClient) queueName message =
        async {
            let sender =
                client.CreateSender(ServiceBusQueueName.toString queueName)

            let! cancellationToken = Async.CancellationToken

            do!
                sender.SendMessageAsync(message, cancellationToken)
                |> Async.AwaitTask
        }

    let batchSendMessages (client: ServiceBusClient) queueName messages =
        let sendBatch (sender: ServiceBusSender) (batch: ServiceBusMessageBatch) =
            async {
                let! cancellationToken = Async.CancellationToken

                do!
                    sender.SendMessagesAsync(batch, cancellationToken)
                    |> Async.AwaitTask
            }

        let createBatch (sender: ServiceBusSender) =
            async {
                let! cancellationToken = Async.CancellationToken

                return!
                    sender.CreateMessageBatchAsync(cancellationToken)
                    |> Async.fromValueTaskOf
            }

        let rec folder (sender: ServiceBusSender) (batch: ServiceBusMessageBatch) message =
            async {
                if batch.TryAddMessage(message) then
                    return batch
                else if batch.Count > 0 then
                    do! sendBatch sender batch

                    let! newBatch = createBatch sender
                    return! folder sender newBatch message
                else
                    return invalidOp "Could not add message to batch. Message might be too big."
            }

        async {
            let sender =
                client.CreateSender(ServiceBusQueueName.toString queueName)

            let! originalBatch = createBatch sender

            let! finalBatch =
                messages
                |> AsyncSeq.ofSeq
                |> AsyncSeq.foldAsync (folder sender) originalBatch

            do! sendBatch sender finalBatch
        }

[<RequireQualifiedAccess>]
module ServiceBusMessageActions =
    let completeMessage (actions: ServiceBusMessageActions) message =
        async {
            let! cancellationToken = Async.CancellationToken

            do!
                actions.CompleteMessageAsync(message, cancellationToken)
                |> Async.AwaitTask
        }

    let abandonMessage (actions: ServiceBusMessageActions) (properties: Map<string, obj>) message =
        async {
            let! cancellationToken = Async.CancellationToken

            do!
                actions.AbandonMessageAsync(message, properties, cancellationToken)
                |> Async.AwaitTask
        }

    let deadLetterMessage (actions: ServiceBusMessageActions) (properties: Map<string, obj>) message =
        async {
            let! cancellationToken = Async.CancellationToken

            do!
                actions.DeadLetterMessageAsync(message, properties, cancellationToken)
                |> Async.AwaitTask
        }
